### DISCUSSION TITLE
Including environmental variables when the spatial coverage is not complete

### DISCUSSION THREAD

**Discussion by ArnoGitHub** [inlabru-org/inlabru#103] [Q&A] [answered]
Hi,

I am currently following a very interesting course showing how we can use inlabru to model animal density distribution based on distance sampling surveys with the potential of adding covariates such as environmental variables. However, it was mentioned that each covariate should be provided as a SpatialPixelsDataFrame with 'complete' data, i.e. with non-missing value at every point within the region defined by the mesh boundary.
Sometimes, in my case, the value of the environmental variable considered (say SST) is available all along my transects and the field of view of the observers, but not for the whole region defined by the mesh boundary (due to cloud cover for example).

I see the point of having a full spatial coverage of the environmental variable of interest for the prediction part, but is it absolutely necessary to the fitting part?

If INLA absolutely needs a full coverage with no NA values but will in fact not use the values outside of the transects footprints for the calculations, can we just fill those NAs with a fake value such as a mean value? Or will it breaks everything?

Thanks in advance

Arnaud

**Comment by finnlindgren** [comment, marked as answer]
As you rightly note, for estimation the covariate only needs to be available in the vicinity of the sampling effort, so for fully serveyed regions that's the whole domain, but for transect data it's in the vicinity of the transects. The size of the needed vicinity is determined by the resolution of the mesh.
If the covariate is not available (or evaluates to NA) at all the needed locations, and the covariate is stored as a Spatial/Pixels/Grid/DataFrame, inlabru will call the internal (not currently exported) inlabru:::bru_fill_missing() function to do automatic infilling based on the nearest known values.  There is unfortunately no automatic method for associating covariate values directly with the transects themselves; inlabru only knows how to do it from Spatial*DataFrame inputs.

In short: As long as you supply data on a grid near the transects, it should be ok for estimation, but you may get a warning that some local infilling is happening.

**Reply by finnlindgren** [reply]
In the next release, the `bru_fill_missing` function will be exported as well. You can see the documentation at [https://github.com/inlabru-org/inlabru/blob/devel/man/bru_fill_missing.Rd](https://github.com/inlabru-org/inlabru/blob/devel/man/bru_fill_missing.Rd), but it doesn't discuss exactly how it's used internally.

**Comment by ArnoGitHub** [comment]
Thanks for this quick reply.
So if I understand correctly, if I provide a SpatialPixelsDataFrame with correct values for my environmental variable (no NAs) in an area sufficiently large around my transects (i.e. large enough to cover the nodes of the mesh in the vicinity of the transects), infilling will not be needed and values outside of the sampled area (i.e. far from the transect) will not be used for the calculations and thus can be NAs.
Am I correct ? 

Thanks for the useful link too !

**Reply by finnlindgren** [reply]
Correct!

**Reply by adietzel** [reply]
Hi,
I know this is an old discussion thread but my question pertains to the same issue of data completeness when fitting an lgcp model with inlabru. 

My covariate data is stored in a SpatialPolygonsDataFrame. All observations are contained within this SPDF but many lie close to its edge. I would like to construct an spde mesh with inner boundary defined by the SPDF. I wanted to also include an outer boundary in the mesh to address boundary effects but no covariate data is available beyond the inner boundary. When I construct a mesh with inner and outer boundary and fit the lgcp model, I receive the warning that the model input contains NA values and that inlabru is attempting to fill them in spatially by nearest available value. I presume this means that covariate data needs to be available for the entire mesh if infilling is to be avoided, and that I am faced with the choice of either having no outer boundary at all or creating an inner boundary within distance x of my outer boundary and thus losing all observations that fall between the two boundaries? Or is there a sensible way to fit an lgcp model without having covariate data between the inner and outer boundary of the mesh while avoiding infilling? If covariate data needs to span the entire mesh, I presume losing observations is the better choice than having no outer boundary in the mesh?

I am new to using INLA and inlabru and building spde meshes and would appreciate your advice on this. 
Thanks a lot for your help. 
Best

Andreas

**Reply by finnlindgren** [reply]
@adietzel, as long as your covariate is available throughout your study domain (i.e. the polygon defining your sampler information), the automatic infill is only there to ensure that there are no accidental NAs at the edge of the domain. The likelihood wouldn't actually use the infilled values, but inlabru can't know that in advance, so it fills in all the values to make sure.

**Reply by adietzel** [reply]
Great, thanks for your quick response. From your experience, will the infilling slow down the model fit substantially? It does seem to take a while longer. Thanks

**Reply by finnlindgren** [reply]
You could fill in the grid before running the model with the help of `inlabru::bru_fill_missing()` that inlabru uses internally, but the automatic infill normally only needs to run once, so this only helps if you store the infilled version for later use.

**Reply by adietzel** [reply]
Great tip, thanks. I am running the same script for 100+ species so doing the infilling prior to model fitting should save a lot of time. 
Thanks for your help!

**Reply by finnlindgren** [reply]
You'll need something like this (I've assumed the grid data frame column is called "values":
```r
# Set target grid parameters to create a grid covering the needed region/mesh;
#   likely just a complete grid version of the source_grid:
target_grid <- SpatialGridDataFrame(...)
target_values <- bru_fill_missing(source_grid, target_grid, target_grid@data$values, layer = "values")
target_grid@data$values <- target_values
```
There isn't currently a wrapper for this use case, so you'll need the manual approach shown here.
Since we're implementing sf/terra/etc support, and sp/rgdal are being retired by the maintainters, when we do make a wrapper, it will likely involve the new packages instead of sp.

**Reply by adietzel** [reply]
Thanks for the code snippet, Finn! If the likelihood is indeed not using the values in the grid that fall outside the domain/samplers region, could I also just expand the grid manually and fill in the outer cells with zeros or an arbitrary value? Or would you recommend always going the bru_fill_missing route? Cheers

**Comment by finnlindgren** [comment]
It needs proper values in the vicinity of the domain so zeros is
inappropriate. Fill_missing is what you need.

**Comment by adietzel** [comment]
Excellent. Thank you!

Sent from my iPhone

On 19 Mar 2022, at 6:24 am, Finn Lindgren ***@***.***> wrote:

﻿

It needs proper values in the vicinity of the domain so zeros is
inappropriate. Fill_missing is what you need.

—
Reply to this email directly, view it on GitHub<https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Finlabru-org%2Finlabru%2Fdiscussions%2F103%23discussioncomment-2389962&data=04%7C01%7Candreas.dietzel%40my.jcu.edu.au%7Cabc8a864a8864e785c6a08da0914e3d9%7C2eba4cf8af764db3bcaf81b5592535ef%7C0%7C0%7C637832282575365516%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=BeN%2BphlbT6ErARiZSv%2FRPiznNkYn%2FlbKL3Hu9y8Xnmw%3D&reserved=0>, or unsubscribe<https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAOYIJJIJONSWDHTRHWCYQ5TVATJ5ZANCNFSM4X26ZMBA&data=04%7C01%7Candreas.dietzel%40my.jcu.edu.au%7Cabc8a864a8864e785c6a08da0914e3d9%7C2eba4cf8af764db3bcaf81b5592535ef%7C0%7C0%7C637832282575365516%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=jIld4ksknZkvqc5UG6Sosqd8mTmLzySLL97JsnSW0xY%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Candreas.dietzel%40my.jcu.edu.au%7Cabc8a864a8864e785c6a08da0914e3d9%7C2eba4cf8af764db3bcaf81b5592535ef%7C0%7C0%7C637832282575365516%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=KhFvG78GHE8u41%2BFqiw80LzTc5JAivd1eaLSNdLDcQs%3D&reserved=0> or Android<https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Candreas.dietzel%40my.jcu.edu.au%7Cabc8a864a8864e785c6a08da0914e3d9%7C2eba4cf8af764db3bcaf81b5592535ef%7C0%7C0%7C637832282575365516%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=X4NjJ6%2B5Bosibd55nseWU%2BbzeAbk8LoXZuzWocSu8Ng%3D&reserved=0>.
